### PR TITLE
Separate core, kernel and third party objects.

### PR DESCRIPTION
**Pull request by njeffrie**
Add a second optimization level in the Makefile to enable different
levels between kernels and the rest of the TFLM code. This results in a
smaller binary with minimal performance impact compared with using a
single optimization level.

Allows the use of implicit patterns to compile all sources, choosing
different flags for core, kernel and third party sources.

The following measurements are taken using the hexagon toolchain +
hexagon-size and hexagon-sim.

For the keyword benchmark using -O2:
   text    data     bss     dec
  58140   37639   46612  142391

  Cycles: 1700364

For the keyword benchmark using -O2 for kernels and -Oz for framework:
   text    data     bss     dec
  52796   37623   46612  137031

  Cycles: 1759664

### CODE REVIEW

File: tensorflow/lite/micro/tools/make/Makefile [hunk 1]
```diff
@@ -113,7 +113,15 @@ ifneq ($(CO_PROCESSOR),)
...
 endif
 
-OPTIMIZATION_LEVEL := -O3
+CORE_OPTIMIZATION_LEVEL := -Os
+KERNEL_OPTIMIZATION_LEVEL := -O2
+
+# Warn if deprecated optimization level is set.
+OPTIMIZATION_LEVEL :=
+ifneq ($(OPTIMIZATION_LEVEL),)
+echo "OPTIMIZATION_LEVEL is no longer used."
```
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/Makefile:122]: Let's make this an error. Warnings will get lost in the output

**Review by advaitjain — CHANGES_REQUESTED** [review]
One nit. And that will also force your to update the test_ scripts.

File: tensorflow/lite/micro/tools/make/Makefile (hunk 1 above)
Comment (njeffrie) [reply tensorflow/lite/micro/tools/make/Makefile]: Done.

**Review by advaitjain — APPROVED** [review]

File: tensorflow/lite/micro/tools/ci_build/test_bluepill.sh [hunk 2]
```diff
@@ -34,12 +34,12 @@ readable_run make -f tensorflow/lite/micro/tools/make/Makefile TARGET=${TARGET}
 
 # check that the release build is ok.
 readable_run make -f tensorflow/lite/micro/tools/make/Makefile clean
-readable_run make -j8 -f tensorflow/lite/micro/tools/make/Makefile TARGET=${TARGET} OPTIMIZATION_LEVEL=-O3 BUILD_TYPE=release build
+readable_run make -j8 -f tensorflow/lite/micro/tools/make/Makefile TARGET=${TARGET} BUILD_TYPE=release build
 
 # Next, build w/o release so that we can run the tests and get additional
 # debugging info on failures.
 readable_run make -f tensorflow/lite/micro/tools/make/Makefile clean
-readable_run make -j8 -f tensorflow/lite/micro/tools/make/Makefile TARGET=${TARGET} OPTIMIZATION_LEVEL=-Os test
```
Comment (advaitjain) [inline tensorflow/lite/micro/tools/ci_build/test_bluepill.sh:42]: This line is now doing the same as then next one. Let's remove one of them. I'm ok switching bluepill to only use the default optimization levels.

Comment (njeffrie) [reply tensorflow/lite/micro/tools/ci_build/test_bluepill.sh:42]: Done.

**Review by advaitjain — DISMISSED** [review]

**Review by advaitjain — DISMISSED** [review]

**Review by advaitjain — APPROVED** [review]

**Review by advaitjain — APPROVED** [review]

**Pull request merged**